crypto/evp/p_lib.c

```diff
@@ -84,14 +84,34 @@ int EVP_PKEY_save_parameters(EVP_PKEY *pkey, int mode)
     return 0;
 }
 
+/*
+ * TODO(3.0) In case only one of the keys has a provider side "origin", the
+ * other one gets upgraded to have a provider side "origin" as well.  That
+ * means that it drops the pointers to any legacy "origin" is may have.  We
+ * should investigate if that's too aggressive and make such a scenario
+ * unsupported instead.
+ */
 int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from)
 {
-    if (to->type == EVP_PKEY_NONE) {
-        if (EVP_PKEY_set_type(to, from->type) == 0)
-            return 0;
-    } else if (to->type != from->type) {
-        EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_KEY_TYPES);
-        goto err;
+    /*
+     * TODO: clean up legacy stuff from this function when legacy support
+     * is gone.
+     */
+
+    /*
+     * Only check that type match this early when both keys are legacy.
+     * If either of them is provided, we let evp_keymgmt_util_copy()
+     * do this check, after having exported either of them that isn't
+     * provided.
+     */
+    if (!(to->keymgmt != NULL || from->keymgmt != NULL)) {
```

Comment:
I think this is easier..
if ( x == NULL && y == NULL)  {

Reply:
Changed.
(there was good reason for this form, but that reason is gone now, so...)